File: FBDeviceControl/Commands/FBDeviceLogCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceLogCommands.h"

#import "FBAMDServiceConnection.h"
#import "FBDevice.h"
#import "FBDeviceControlError.h"

#pragma mark Protocol Adaptor

@interface FBDeviceLogOperation : NSObject <FBLogOperation>

@property (nonatomic, readonly, strong) FBFileReader *reader;
@property (nonatomic, readonly, strong) FBFuture<NSNull *> *readCompleted;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNull *> *serviceCompleted;

@end

@implementation FBDeviceLogOperation

@synthesize consumer = _consumer;

- (instancetype)initWithConsumer:(id<FBDataConsumer>)consumer readCompleted:(FBFuture<NSNull *> *)readCompleted serviceCompleted:(FBMutableFuture<NSNull *> *)serviceCompleted
{
  self = [self init];
  if (!self) {
    return nil;
  }

  _consumer = consumer;
  _readCompleted = readCompleted;
  _serviceCompleted = serviceCompleted;

  return self;
}

- (FBFuture<NSNull *> *)completed
{
  return self.serviceCompleted;
}

@end

#pragma mark FBDeviceLogCommands Implementation

@interface FBDeviceLogCommands ()

@property (nonatomic, readonly, weak) FBDevice *device;

@end

@implementation FBDeviceLogCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBDevice *)target
{
  return [[self alloc] initWithDevice:target];
}

- (instancetype)initWithDevice:(FBDevice *)device
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;

  return self;
}

#pragma mark FBLogCommands Implementation

- (FBFuture<id<FBLogOperation>> *)tailLog:(NSArray<NSString *> *)arguments consumer:(id<FBDataConsumer>)consumer
{
  if (arguments.count != 0) {
    NSString *unsupportedArgumentsMessage = [NSString stringWithFormat:@"[FBDeviceLogCommands][rdar://38452839] Unsupported arguments: %@", arguments];
    [consumer consumeData:[unsupportedArgumentsMessage dataUsingEncoding:NSUTF8StringEncoding]];
    [self.device.logger log:unsupportedArgumentsMessage];
  }
  dispatch_queue_t queue = self.device.asyncQueue;
  dispatch_queue_t readQueue = dispatch_queue_create("com.facebook.fbdevicecontrol.device_log_consumer", DISPATCH_QUEUE_SERIAL);
  return [[[self.device
            startService:@"com.apple.syslog_relay"]
           onQueue:queue
           pend:^(FBAMDServiceConnection *connection) {
             id<FBFileReaderProtocol> reader = [connection readFromConnectionWritingToConsumer:consumer onQueue:readQueue];
             return [[reader startReading] mapReplace:reader];
           }]
          onQueue:queue
          enter:^(id<FBFileReaderProtocol> reader, FBMutableFuture<NSNull *> *teardown) {
            FBFuture<NSNull *> *readCompleted = [[reader finishedReading] mapReplace:NSNull.null];
            return [[FBDeviceLogOperation alloc] initWithConsumer:consumer readCompleted:readCompleted serviceCompleted:teardown];
          }];
}

@end

File: FBDeviceControl/Commands/FBDeviceLogCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

// MARK: - FBDeviceLogOperation

@objc(FBDeviceLogOperation)
public class FBDeviceLogOperation: NSObject, FBLogOperation {
  public let consumer: any FBDataConsumer
  private let readCompleted: FBFuture<NSNull>
  private let serviceCompleted: FBMutableFuture<NSNull>

  // MARK: Initializers

  init(
    consumer: any FBDataConsumer,
    readCompleted: FBFuture<NSNull>,
    serviceCompleted: FBMutableFuture<NSNull>
  ) {
    self.consumer = consumer
    self.readCompleted = readCompleted
    self.serviceCompleted = serviceCompleted
    super.init()
  }

  // MARK: FBiOSTargetOperation

  public var completed: FBFuture<NSNull> {
    return unsafeBitCast(serviceCompleted, to: FBFuture<NSNull>.self)
  }
}

// MARK: - FBDeviceLogCommands

@objc(FBDeviceLogCommands)
public class FBDeviceLogCommands: NSObject, FBLogCommands {
  private weak var device: FBDevice?

  // MARK: - Initializers

  public class func commands(with target: any FBiOSTarget) -> Self {
    return self.init(device: target as! FBDevice)
  }

  required init(device: FBDevice) {
    self.device = device
    super.init()
  }

  // MARK: - FBLogCommands

  public func tailLog(_ arguments: [String], consumer: any FBDataConsumer) -> FBFuture<any FBLogOperation> {
    guard let device else {
      return FBFuture(error: FBDeviceControlError().describe("Device is nil").build())
    }
    if !arguments.isEmpty {
      let unsupportedArgumentsMessage = "[FBDeviceLogCommands][rdar://38452839] Unsupported arguments: \(arguments)"
      if let data = unsupportedArgumentsMessage.data(using: .utf8) {
        consumer.consumeData(data)
      }
      device.logger?.log(unsupportedArgumentsMessage)
    }
    let queue = device.asyncQueue
    let readQueue = DispatchQueue(label: "com.facebook.fbdevicecontrol.device_log_consumer")
    return
      device
      .startService("com.apple.syslog_relay")
      .onQueue(
        queue,
        enter: { connection, teardown -> Any in
          let reader = connection.readFromConnectionWriting(to: consumer, on: readQueue)
          reader.startReading()
          let readCompleted = reader.finishedReading(withTimeout: .infinity).mapReplace(NSNull()) as! FBFuture<NSNull>
          return FBDeviceLogOperation(
            consumer: consumer,
            readCompleted: readCompleted,
            serviceCompleted: teardown
          )
        }) as! FBFuture<any FBLogOperation>
  }
}

// MARK: - FBDevice+AsyncLogCommands

extension FBDevice: AsyncLogCommands {

  public func tailLog(arguments: [String], consumer: any FBDataConsumer) async throws -> any AsyncLogOperation {
    let operation = try await bridgeFBFuture(logCommands().tailLog(arguments, consumer: consumer))
    return AsyncLogOperationBridge(operation)
  }
}
